crates/da_clients/ethereum/src/lib.rs

```diff
@@ -1,38 +1,212 @@
 #![allow(missing_docs)]
 #![allow(clippy::missing_docs_in_private_items)]
+use alloy::consensus::{
+    BlobTransactionSidecar, SignableTransaction, TxEip4844, TxEip4844Variant, TxEip4844WithSidecar, TxEnvelope,
+};
+use alloy::eips::{eip2718::Encodable2718, eip2930::AccessList, eip4844::BYTES_PER_BLOB};
+use alloy::network::{Ethereum, TxSigner};
+use alloy::primitives::{bytes, FixedBytes, TxHash, U256, U64};
+use alloy::providers::{Provider, ProviderBuilder, RootProvider};
 use alloy::rpc::client::RpcClient;
+use alloy::signers::wallet::LocalWallet;
 use alloy::transports::http::Http;
 use async_trait::async_trait;
+
 use color_eyre::Result;
+use mockall::{automock, predicate::*};
 use reqwest::Client;
-use starknet::core::types::FieldElement;
 use std::str::FromStr;
 use url::Url;
 
+use c_kzg::{Blob, KzgCommitment, KzgProof, KzgSettings};
 use config::EthereumDaConfig;
 use da_client_interface::{DaClient, DaVerificationStatus};
-
+use dotenv::dotenv;
+use std::{env, path::Path};
 pub mod config;
 pub struct EthereumDaClient {
     #[allow(dead_code)]
-    provider: RpcClient<Http<Client>>,
+    provider: RootProvider<Ethereum, Http<Client>>,
+    wallet: LocalWallet,
+    trusted_setup: KzgSettings,
 }
 
+#[automock]
 #[async_trait]
 impl DaClient for EthereumDaClient {
-    async fn publish_state_diff(&self, _state_diff: Vec<FieldElement>) -> Result<String> {
-        unimplemented!()
+    async fn publish_state_diff(&self, state_diff: Vec<Vec<u8>>) -> Result<String> {
+        dotenv().ok();
+        let provider = &self.provider;
+        let trusted_setup = &self.trusted_setup;
+        let wallet = &self.wallet;
+        let addr = wallet.address();
+
+        let (sidecar_blobs, sidecar_commitments, sidecar_proofs) = prepare_sidecar(&state_diff, trusted_setup).await?;
+        let sidecar = BlobTransactionSidecar::new(sidecar_blobs, sidecar_commitments, sidecar_proofs);
+
+        let eip1559_est = provider.estimate_eip1559_fees(None).await?;
+        let chain_id: u64 = provider.get_chain_id().await?.to_string().parse()?;
+
+        let max_fee_per_blob_gas: u128 = provider.get_blob_base_fee().await?.to_string().parse()?;
+        let max_priority_fee_per_gas: u128 = provider.get_max_priority_fee_per_gas().await?.to_string().parse()?;
+
+        let tx = TxEip4844 {
+            chain_id,
+            nonce: 0, // can be block number
```

Comment:
If I'm not missing something, it would not be set right?
There should be NonceFiller in the alloy provider chain to handle that, couldn't find it explicitly set.

Reply:
ya makes sense, @Mohiiit let's fetch the nonce from the RPC itself